### PR TITLE
Fix redundant code checks

### PR DESCRIPTION
**Pull request by wjn740**
perlcritic,test-no-wait_idle phony targets have provided
code check. remove these actions in 'test-merge' also could give help
to performance.
Move test-merge target to the end, so that start 'merge'
after all checks are PASSED.

#7775  be blocked by this issue.

### CODE REVIEW

**Comment by okurz** [comment]
@coolo I never properly understood the "test-merge" part. Can you please review?

**Comment by coolo** [comment]
Well, push a wait_idle call to the branch and see if it still fails.

**Comment by wjn740** [comment]
https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/554148357

@coolo "wait_idle" warning works.

**Comment by okurz** [comment]
ok, cool. Why do we even need the whole make target then? :confused:

**Comment by wjn740** [comment]
@okurz I don't know might be it could be extended in future.

**Comment by wjn740** [comment]
If decided to remove it, no problem, I could modify it.

**Comment by coolo** [comment]
the idea back then was to be quick about it - now we seem to run all tests all the time anyway. So just remove it

**Comment by wjn740** [comment]
https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/554486999
after emove "test-merge", the "wait-idle" still be reported.

**Comment by wjn740** [comment]
squash all commits into one.

File: Makefile [hunk 1]
```diff
@@ -63,19 +63,6 @@ test-metadata:
 test-metadata-changed:
 	tools/check_metadata $$(git diff --name-only | grep 'tests.*pm')
 
-.PHONY: test-merge
-test-merge:
-	@REV=$$(git merge-base FETCH_HEAD master 2>/dev/null) ;\
-	if test -n "$$REV"; then \
-	  FILES=$$(git diff --name-only FETCH_HEAD `git merge-base FETCH_HEAD master 2>/dev/null` | grep 'tests.*pm') ;\
-	  for file in $$FILES; do if test -f $$file; then \
-	    tools/check_metadata $$file || touch failed; \
```
Comment (coolo) [inline Makefile]: So what checks the metadata now?

Comment (wjn740) [reply Makefile]: missing, I add it now.

**Comment by wjn740** [comment]
```
tools/check_metadata $(git diff --name-only | grep 'tests.*pm')
tools/check_metadata: 2: tools/check_metadata: @: Need files to check as argument
make: *** [test-metadata-changed] Error 2
The command "make test" exited with 2.
```
It seems like we need remove the limitation of check_metadata.

**Comment by wjn740** [comment]
@coolo instead remove checking in check_metadata, I modify Makefile, only call check_metadata script when there some ".pm" file be committed.

**Comment by okurz** [comment]
Please wait a second. I still think the idea of "test-merge" is a good one. The runtime of tests for os-autoinst-distri-opensuse increased over the past months and "test-merge" is a good approach to counter this, especially for perl critic. Can we rather try to do it correctly again? I am not sure which commit started the mess – could very well be one of mine :)

File: Makefile [hunk 2]
```diff
@@ -89,7 +76,7 @@ test-spec:
 	tools/update_spec --check
 
 .PHONY: test-static
-test-static: tidy-check test-yaml-valid test-merge test-dry test-no-wait_idle test-unused-modules test-soft_failure-no-reference test-spec test-invalid-syntax
+test-static: tidy-check test-yaml-valid test-metadata-changed test-dry test-no-wait_idle test-unused-modules test-soft_failure-no-reference test-spec test-invalid-syntax
```
Comment (okurz) [inline Makefile]: this is wrong because `test-metadata-changed` is supposed to be called only for uncommited files but `test-static` is supposed to be called on a "commited state" instead like all the other checks.

Comment (wjn740) [reply Makefile]: OK, let me try again.

**Comment by wjn740** [comment]
https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/554596248

@coolo  @okurz Please review this.

The original issue is CI report "test-merge" failed, without any information.
and the failure happened at 
`"git grep -q "wait-idle" $$file"` && 
due to "-q" make silence so user totally doesn't know what happens. 

Might be we could change something here.

File: Makefile [hunk 3]
```diff
@@ -69,12 +69,13 @@ test-merge:
 	if test -n "$$REV"; then \
 	  FILES=$$(git diff --name-only FETCH_HEAD `git merge-base FETCH_HEAD master 2>/dev/null` | grep 'tests.*pm') ;\
 	  for file in $$FILES; do if test -f $$file; then \
-	    tools/check_metadata $$file || touch failed; \
-	    git grep -q wait_idle $$file && touch failed; \
-	    ${PERLCRITIC} $$file || (echo $$file ; touch failed) ;\
+	    tools/check_metadata $$file || touch failed1; \
+	    git grep -q wait_idle $$file && echo $$file >>failed2; \
```
Comment (okurz) [inline Makefile]: I guess you can simply output the file here instead of delaying that (in the rather complicated way) to later.

Would be great if we could simply reuse the make targets as defined above with the "-changed" suffix, e.g. as an optional argument. If "reusing a make target" is not possible we can define the commands within a Makefile variable and use the command here and in the make targets that check all the files.

Comment (wjn740) [reply Makefile]: First, if I simply output the file, users might can't understand what mistake happen. 

About reuse the make targets as defined above with the "changed" suffix,
check_metadata script needs an argument so line 62~64 is wrong. because when user commits modification on !".pm" files, this checking go failed.  but here, due to FOR loop will filter NULL , so we needn't care here.

Comment (wjn740) [reply Makefile]: Our checking only working for ".pm" files ? If we also care !".pm" files, if we ONLY care ".pm" files, I could modify "check_metadata" script turn to an "optional argument".

Comment (okurz) [reply Makefile]: yes, it's ok if the "check_metadata" is only executed for "\.pm" files, actually even only "tests/.*\.pm" would be enough.

**Comment by wjn740** [comment]
I think about this change, actually, the "wait_idle checking" can't report error content is the root cause.

Keep the commit clear/simple, we don't remove test-merge now, just modify the line of "wait_idle". make it can report something like "xxxxxxx file include "wait_idle" please remove".

So, there are three solutions, 
add a script in tools, seems like tools/test_no_wait_idle, to implement a relative complex check and report.
another one, just modify this line to some like "(echo "no_wait_idle check"; echo $$file; touch failed;)"

Three, just remove "wait_idle" line in 'test-merge'. that be followed by 'test-no-wait_idle'. This solution is just slow inefficiently.

File: Makefile (hunk 3 above)
Comment (wjn740) [reply Makefile]: @okurz please review the following comments.

**Comment by okurz** [comment]
yes, sure. I agree with your second solution, that is: 'just modify this line to some like "(echo "no_wait_idle check"; echo $$file; touch failed;)"'

**Comment by wjn740** [comment]
> yes, sure. I agree with your second solution, that is: 'just modify this line to some like "(echo "no_wait_idle check"; echo $$file; touch failed;)"'

Patch is on the way.

**Comment by wjn740** [comment]
the latest check code example:
https://travis-ci.org/os-autoinst/os-autoinst-distri-opensuse/jobs/556602280

I will squash branch ready to merge.

**Comment by wjn740** [comment]
This PR make sure CI could give prompt when the code has some of the mistakes, about remove the redudant code, I will file another PR, after thoughtful consideration.

**Comment by wjn740** [comment]
@okurz , ready to merge.

File: Makefile [hunk 4]
```diff
@@ -70,7 +70,7 @@ test-merge:
 	  FILES=$$(git diff --name-only FETCH_HEAD `git merge-base FETCH_HEAD master 2>/dev/null` | grep 'tests.*pm') ;\
 	  for file in $$FILES; do if test -f $$file; then \
 	    tools/check_metadata $$file || touch failed; \
-	    git grep -q wait_idle $$file && touch failed; \
+	    git --no-pager grep wait_idle $$file ||  touch failed; \
```
Comment (okurz) [inline Makefile]: why are you turning around the logic? Grep succeeds when it *finds* a match.

Comment (wjn740) [reply Makefile]: I think it's a typo. Sorry.

**Comment by okurz** [comment]
Let's get this over with :)

**Comment by wjn740** [comment]
ok, i will continuly improve.
